File: producer/lib/sqs.js
'use strict';

const { SQSClient, SendMessageBatchCommand } = require('@aws-sdk/client-sqs');

const client = new SQSClient({ region: process.env.AWS_REGION || 'eu-west-1' });

const SQS_BATCH_SIZE = 10; // limite AWS fisso per sendMessageBatch

/**
 * Pubblica un array di record su SQS in batch da 10.
 * Lancia un errore se anche un solo messaggio del batch fallisce.
 */
async function publishBatch(records, queueUrl) {
  const chunks = chunkArray(records, SQS_BATCH_SIZE);

  for (const chunk of chunks) {
    const entries = chunk.map((record, i) => ({
      Id: String(i),
      MessageBody: JSON.stringify(record),
    }));

    const response = await client.send(
      new SendMessageBatchCommand({ QueueUrl: queueUrl, Entries: entries })
    );

    if (response.Failed?.length > 0) {
      console.error('[SQS] messaggi falliti nel batch:', JSON.stringify(response.Failed));
      throw new Error(`sendMessageBatch: ${response.Failed.length} messaggi falliti`);
    }
  }
}

function chunkArray(arr, size) {
  const result = [];
  for (let i = 0; i < arr.length; i += size) {
    result.push(arr.slice(i, i + size));
  }
  return result;
}

module.exports = { publishBatch };

File: producer/lib/sqs.py
import json
import uuid
import boto3

_client = boto3.client("sqs", region_name="eu-west-1")


def publish_batch(records: list[dict], queue_url: str) -> int:
    """Publish records to SQS in chunks of 10 (AWS hard limit). Returns total sent."""
    total = 0
    for i in range(0, len(records), 10):
        chunk = records[i : i + 10]
        entries = [
            {"Id": str(uuid.uuid4()), "MessageBody": json.dumps(record)}
            for record in chunk
        ]
        response = _client.send_message_batch(QueueUrl=queue_url, Entries=entries)
        failed = response.get("Failed", [])
        if failed:
            raise RuntimeError(f"SQS batch send partial failure: {failed}")
        total += len(chunk)
    return total
